lib/elixir/lib/calendar/datetime.ex

```diff
@@ -172,20 +172,129 @@ defmodule DateTime do
+      iex> first_dt
+      #DateTime<2018-10-28 02:30:00+02:00 CEST Europe/Copenhagen>
+      iex> second_dt
+      #DateTime<2018-10-28 02:30:00+01:00 CET Europe/Copenhagen>
+
+  When there is a gap in wall time - for instance in spring when the clocks are
+  turned forward - the latest valid datetime just before the gap and the first
+  valid datetime just after the gap.
+
+      iex> {:gap, just_before, just_after} = DateTime.from_naive(~N[2019-03-31 02:30:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> just_before
+      #DateTime<2019-03-31 01:59:59.999999+01:00 CET Europe/Copenhagen>
+      iex> just_after
+      #DateTime<2019-03-31 03:00:00+02:00 CEST Europe/Copenhagen>
+
+  Most of the time there is one, and just one, valid datetime for a certain
+  date and time in a certain time zone.
+
+      iex> {:ok, datetime} = DateTime.from_naive(~N[2018-07-28 12:30:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> datetime
+      #DateTime<2018-07-28 12:30:00+02:00 CEST Europe/Copenhagen>
+
   """
   @doc since: "1.4.0"
-  @spec from_naive(NaiveDateTime.t(), Calendar.time_zone()) :: {:ok, t}
-  def from_naive(naive_datetime, time_zone)
+  @spec from_naive(
+          NaiveDateTime.t(),
+          Calendar.time_zone(),
+          TimeZoneDatabaseClient.tz_db_or_config()
+        ) ::
+          {:ok, t}
+          | {:ambiguous, t, t}
+          | {:gap, t, t}
+          | {:error, :time_zone_not_found}
+          | {:error, :incompatible_calendars}
+          | {:error, :no_time_zone_database}
+
+  def from_naive(naive_datetime, time_zone, tz_db_or_config \\ :from_config)
+
+  def from_naive(naive_datetime, "Etc/UTC", _) do
+    utc_period = %{std_offset: 0, utc_offset: 0, zone_abbr: "UTC"}
+    {:ok, from_naive_with_period(naive_datetime, "Etc/UTC", utc_period)}
+  end
+
+  def from_naive(%{calendar: Calendar.ISO} = naive_datetime, time_zone, tz_db_or_config) do
+    case TimeZoneDatabaseClient.time_zone_periods_from_wall_datetime(
+           naive_datetime,
+           time_zone,
+           tz_db_or_config
+         ) do
+      {:single, period} ->
+        {:ok, from_naive_with_period(naive_datetime, time_zone, period)}
+
+      {:ambiguous, first_period, second_period} ->
+        first_datetime = from_naive_with_period(naive_datetime, time_zone, first_period)
+        second_datetime = from_naive_with_period(naive_datetime, time_zone, second_period)
+        {:ambiguous, first_datetime, second_datetime}
+
+      {:gap, {first_period, first_period_until_wall}, {second_period, second_period_from_wall}} ->
+        # `until_wall` is not valid, but any time just before is.
+        # So by subtracting a second and adding .999999 seconds
+        # we get the last microsecond just before.
+        before_naive =
+          first_period_until_wall
+          |> Map.put(:microsecond, {999_999, 6})
+          |> NaiveDateTime.add(-1)
```

Comment:
I wonder if we should ask the timezone database to return the last valid instant so we can skip this conversion. Thoughts @lau?

Reply:
Documenting this for posterity. José and I just talked about this and agreed to keep the handling here for the following reason:

When the gap starts, the start of the gap is invalid. What we want here is the last possible valid time just before a gap. In this case it is 6 nines (.999999). If the timezone database was to decide it, it would have to know about what precision Elixir is using. Today it is 6 nines. So we would have to change the interface in order to pass a fractional seconds with a lot of nines just for this purpose. And the timezone database or interface would decide how many nines. Instead it is simpler to let Elixir itself decide this.